src/ApacheOrcDotNet.OptimizedReader/Encodings/OptimizedIntegerRLE.cs

```diff
@@ -0,0 +1,285 @@
+    {
+        enum EncodingType { ShortRepeat, Direct, PatchedBase, Delta }
+
+        public static int ReadValues(ref BufferReader reader, bool isSigned, Span<long> outputValues)
+        {
+            var numReadValues = 0;
+
+            if (!reader.TryRead(out var firstByte))
+                return numReadValues;
+
+            var encodingType = (EncodingType)((firstByte >> 6) & 0x3);
+
+            switch (encodingType)
+            {
+                case EncodingType.ShortRepeat:
+                    ReadShortRepeatValues(firstByte, isSigned, ref reader, outputValues, ref numReadValues);
+                    break;
+                case EncodingType.Direct:
+                    ReadDirectValues(firstByte, isSigned, ref reader, outputValues, ref numReadValues);
+                    break;
+                case EncodingType.PatchedBase:
+                    ReadPatchedBaseValues(firstByte, isSigned: false, ref reader, outputValues, ref numReadValues);
+                    break;
+                case EncodingType.Delta:
+                    ReadDeltaValues(firstByte, isSigned, ref reader, outputValues, ref numReadValues);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Invalid encoding type: {encodingType}");
+            };
+
+            return numReadValues;
+        }
+
+        private static int ReadShortRepeatValues(int firstByte, bool isSigned, ref BufferReader reader, Span<long> outputValues, ref int numReadValues)
+        {
+            var width = ((firstByte >> 3) & 0x7) + 1;
+            numReadValues = (firstByte & 0x7) + 3;
+
+            var repeatingValue = ReadLongBE(ref reader, width, isSigned);
+
+            for (int i = 0; i < numReadValues; i++)
+                outputValues[i] = repeatingValue;
+
+            return width;
+        }
+
+        private static void ReadDirectValues(int firstByte, bool isSigned, ref BufferReader reader, Span<long> outputValues, ref int numReadValues)
+        {
+            var encodedWidth = (firstByte >> 1) & 0x1f;
+            var width = encodedWidth.DecodeDirectWidth();
+            numReadValues = (firstByte & 0x1) << 8;
+
+            if (!reader.TryRead(out var nextByte))
+                throw new InvalidOperationException("Read past end of stream");
+
+            numReadValues |= nextByte;
+            numReadValues += 1;
+
+            ReadBitpackedIntegers(ref reader, isSigned, width, numReadValues, outputValues);
+        }
+
+        private static void ReadPatchedBaseValues(int firstByte, bool isSigned, ref BufferReader reader, Span<long> outputValues, ref int numReadValues)
+        {
+            var encodedWidth = (firstByte >> 1) & 0x1f;
+            var directBitWidth = encodedWidth.DecodeDirectWidth();
+            numReadValues = (firstByte & 0x1) << 8;
+
+            Span<byte> nextBytes = stackalloc byte[3];
+            if (!reader.TryCopyTo(nextBytes))
```

Comment:
The performance is pretty much the same. I'll keep as it is, though, to avoid creating and testing multiple variables.

### First benchmark example
|          Method |     Mean |     Error |    StdDev | Rank | Allocated |
|---------------- |---------:|----------:|----------:|-----:|----------:|
|   TryRead3Bytes | 4.260 ns | 0.0830 ns | 0.0776 ns |    1 |         - |
| TryCopyTo3Bytes | 4.253 ns | 0.0496 ns | 0.0440 ns |    1 |         - |


### Second benchmark example
|          Method |     Mean |     Error |    StdDev | Rank | Allocated |
|---------------- |---------:|----------:|----------:|-----:|----------:|
|   TryRead3Bytes | 6.685 ns | 0.0831 ns | 0.0777 ns |    2 |         - |
| TryCopyTo3Bytes | 4.406 ns | 0.0620 ns | 0.0580 ns |    1 |         - |

### Third benchmark example (*)
|          Method |     Mean |     Error |    StdDev | Rank | Allocated |
|---------------- |---------:|----------:|----------:|-----:|----------:|
|   TryRead3Bytes | 4.350 ns | 0.1104 ns | 0.1547 ns |    1 |         - |
| TryCopyTo3Bytes | 4.347 ns | 0.1140 ns | 0.1358 ns |    1 |         - |

** While benchmarking these, the values in the third example are the ones that repeat the most. The second example can be considered an outlier.